### PR TITLE
Python base feature

### PR DESCRIPTION
**Pull request by arturkamieniecki**
I've added support for creating a feature using python. With it I've created an example assistant feature.

I've modified the interfaces for creating the pytohn feature by creating a: `CppFeatureBase` and `PythonFeatureBase`. Both of them inherit from `FeatureBase`.

### CODE REVIEW

File: Gems/GenAIFramework/Code/Include/GenAIFramework/Feature/AIAgentBus.h [hunk 1]
```diff
@@ -31,9 +32,31 @@ namespace GenAIFramework
         //! The response will be sent to the OnAIResponse notification.
         virtual void SendPrompt(const AIMessages& prompt) = 0;
 
+        //! Send a prompt to the model agent as a JSON string.
+        //! @param prompt The prompt to send as a JSON string.
+        //! The prompt is a JSON string that is an array of objects. Each object consists of a role and content.
+        //! The role is a string that can be "user", "assistant", or "system". Corrsponding to the user, assistant, and system roles.
+        //! The content is an array of strings that are the content of the message.
+        virtual void SendPromptAsJsonString(const AZStd::string& prompt)
```
Comment (jhanca-robotecai) [inline Gems/GenAIFramework/Code/Include/GenAIFramework/Feature/AIAgentBus.h:40]: I see a lack of consistency in the naming: `SendPromptAsJsonString` and `GetHistoryAsJson`. We could use the same name for sending: `SendPrompt` and keep `GetHistoryAsJson` as is. Or `SendTextPrompt` and `GetHistoryAsText`. Whatever :)

File: Gems/GenAIFramework/Code/Include/GenAIFramework/Feature/FeatureBase.h
Comment (jhanca-robotecai) [inline Gems/GenAIFramework/Code/Include/GenAIFramework/Feature/FeatureBase.h:1]: Do we need to introduce three layers of inheritance to keep two variables (`m_agentId` and `m_conversationId`)? It looks complicated and I do not think we will ever need to extend the interface of `FeatureBase`.

I would rather keep `FeatureBase` for C++ (this is a default language in O3DE and should be considered a default language for implementing features) and `FeatureBasePython` for the guest programming language.

File: Gems/GenAIFramework/Code/Include/GenAIFramework/Feature/JsonConversionUtils.h [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+#pragma once
+
+#include <GenAIFramework/Communication/AIModelRequestBus.h>
+
+#include <AzCore/JSON/document.h>
+#include <AzCore/JSON/rapidjson.h>
+#include <AzCore/JSON/writer.h>
+#include <rapidjson/rapidjson.h>
+
+namespace GenAIFramework::Internal
```
Comment (jhanca-robotecai) [inline Gems/GenAIFramework/Code/Include/GenAIFramework/Feature/JsonConversionUtils.h:18]: I would use a more meaningful name for the namespace. E.g. `GenAIFramework::JsonUtils` (and the filename: `JsonUtils.h`).

File: Gems/GenAIFramework/Code/Include/GenAIFramework/Feature/JsonConversionUtils.h [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+ */
+
+#pragma once
+
+#include <GenAIFramework/Communication/AIModelRequestBus.h>
+
+#include <AzCore/JSON/document.h>
+#include <AzCore/JSON/rapidjson.h>
+#include <AzCore/JSON/writer.h>
+#include <rapidjson/rapidjson.h>
```
Comment (jhanca-robotecai) [inline Gems/GenAIFramework/Code/Include/GenAIFramework/Feature/JsonConversionUtils.h:16]: Do you need both, `<AzCore/JSON/rapidjson.h>` and `<rapidjson/rapidjson.h>`?

File: Gems/GenAIFramework/Code/Include/GenAIFramework/GenAIFrameworkTypeIds.h [hunk 4]
```diff
@@ -39,6 +39,10 @@ namespace GenAIFramework
     inline constexpr const char* GenAIFrameworkActionRequestsTypeId = "{D2D195A2-C720-4961-9AEF-C00FCB0AA70D}";
     inline constexpr const char* GenAIFrameworkEditorRequestsTypeId = "{B20BA166-2237-4F0C-AFCD-FB17B63A9CF6}";
     inline constexpr const char* FeatureBaseTypeId = "{4050040A-31C3-4949-8CE0-52FAD7562F4D}";
+    inline constexpr const char* CppFeatureBaseTypeId = "{040a5281-e429-49b8-93fd-d76019cd6099}";
+    inline constexpr const char* PythonFeatureBaseTypeId = "{1a248923-79e4-4ec7-b90b-fa81289e95e7}";
+    inline constexpr const char* BehaviorConversationBusHandlerTypeId = "{bbec0750-3d2e-4631-a9fa-c5f2bac44632}";
+    inline constexpr const char* BehaviourAIAgentNotificationsBusHandlerTypeId = "{119148aa-e7a5-4520-97e0-602953969afb}";
```
Comment (jhanca-robotecai) [inline Gems/GenAIFramework/Code/Include/GenAIFramework/GenAIFrameworkTypeIds.h:45]: Please use uppercase for the consistency

File: Gems/GenAIFramework/Code/Source/PythonTestFeature/PythonTestFeature.cpp [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+        const auto pythonAssistantScriptPath = AZ::IO::Path(thisGemPath) / AZ::IO::Path(PythonAssistantScript);
+        m_assistantPythonScriptLocation = pythonAssistantScriptPath.String();
+
+        bool isOk = false;
+        AZStd::string_view agentIdString = AZStd::to_string(m_conversationId);
+        AZStd::string_view conversationIdString = AZStd::to_string(m_conversationId);
+
+        std::cout << "Executing Python script with agentId: " << agentIdString.cbegin()
+                  << " and conversationId: " << conversationIdString.cbegin() << std::endl;
```
Comment (jhanca-robotecai) [inline Gems/GenAIFramework/Code/Source/PythonTestFeature/PythonTestFeature.cpp:40]: Please remember to clean all debug messages before making this draft ready for the review

File: Gems/GenAIFramework/Code/Source/PythonTestFeature/PythonTestFeature.cpp [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    void PythonAssistantFeature::Reflect(AZ::ReflectContext* context)
+    {
+        if (auto* serializeContext = azrtti_cast<AZ::SerializeContext*>(context))
+        {
+            serializeContext->Class<PythonAssistantFeature, PythonFeatureBase>()->Version(0);
+        }
+
+        if (auto registrationContext = azrtti_cast<GenAIFramework::SystemRegistrationContext*>(context))
+        {
+            registrationContext->RegisterFeature<PythonAssistantFeature>("O3DE Python assistant feature");
```
Comment (jhanca-robotecai) [inline Gems/GenAIFramework/Code/Source/PythonTestFeature/PythonTestFeature.cpp:61]: Please rename `PythonTestFeature` -> `O3DEAssistantPythonFeature` or `O3DEAssistantFeaturePython`
```suggestion
            registrationContext->RegisterFeature<PythonAssistantFeature>("O3DE Assistant (Python)");
```

File: Gems/GenAIFramework/Editor/Scripts/assistant.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
+# assumes a level has been opened or created
```
Comment (jhanca-robotecai) [inline Gems/GenAIFramework/Editor/Scripts/assistant.py:1]: Missing license

**Review by jhanca-robotecai — COMMENTED** [review]
Some comments on the first draft. I will review and test when ready.

File: Gems/GenAIFramework/Code/Include/GenAIFramework/Feature/AIAgentBus.h (hunk 1 above)
Comment (arturkamieniecki) [reply Gems/GenAIFramework/Code/Include/GenAIFramework/Feature/AIAgentBus.h:40]: I would keep the `SendPromptAsJsonString` name as O3DE EBus calling seems not to support function overloading.

File: Gems/GenAIFramework/Code/Include/GenAIFramework/Feature/FeatureBase.h
Comment (arturkamieniecki) [reply Gems/GenAIFramework/Code/Include/GenAIFramework/Feature/FeatureBase.h:1]: Resolved in offline disscusion.

File: Gems/GenAIFramework/Code/Include/GenAIFramework/Feature/JsonUtils.h [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+#include <AzCore/JSON/rapidjson.h>
+#include <AzCore/JSON/writer.h>
+
+namespace GenAIFramework::JsonUtils
+{
+    inline rapidjson::Value AIMessageToJsonValue(const AIMessage& message, rapidjson::Document::AllocatorType& allocator)
+    {
+        rapidjson::Value AIMessageJson(rapidjson::kObjectType);
+
+        rapidjson::Value entryJson(rapidjson::kObjectType);
```
Comment (jhanca-robotecai) [inline Gems/GenAIFramework/Code/Include/GenAIFramework/Feature/JsonUtils.h:23]: Unused variable
```suggestion
```

File: Gems/GenAIFramework/Code/Include/GenAIFramework/Feature/JsonUtils.h [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+
+        AZStd::string role;
+
```
Comment (jhanca-robotecai) [inline Gems/GenAIFramework/Code/Include/GenAIFramework/Feature/JsonUtils.h:26]:
```suggestion
```

File: Gems/GenAIFramework/Code/Include/GenAIFramework/Feature/JsonUtils.h [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+        rapidjson::Value contentJson(rapidjson::kArrayType);
+        for (const auto& content : message.second)
+        {
+            rapidjson::Value contentValue;
+            if (content.type() == AZ::AzTypeInfo<AZStd::string>().Uuid())
+            {
+                contentValue.SetString(AZStd::any_cast<AZStd::string>(content).c_str(), allocator);
+            }
+
+            contentJson.PushBack(contentValue, allocator);
```
Comment (jhanca-robotecai) [inline Gems/GenAIFramework/Code/Include/GenAIFramework/Feature/JsonUtils.h:53]: Does it make sense to push empty values?
```suggestion
            if (content.type() == AZ::AzTypeInfo<AZStd::string>().Uuid())
            {
                rapidjson::Value contentValue;
                contentValue.SetString(AZStd::any_cast<AZStd::string>(content).c_str(), allocator);
                contentJson.PushBack(contentValue, allocator);
            }
```

File: Gems/GenAIFramework/Code/Include/GenAIFramework/Feature/JsonUtils.h [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+        case Role::Assistant:
+            role = "assistant";
+            break;
+        case Role::System:
+            role = "system";
+            break;
+        }
+
+        rapidjson::Value roleValue;
+        roleValue.SetString(role.c_str(), allocator);
```
Comment (jhanca-robotecai) [inline Gems/GenAIFramework/Code/Include/GenAIFramework/Feature/JsonUtils.h:41]: No need for `role` string - you might set the value directly in _switch/case_

File: Gems/GenAIFramework/Code/Source/O3DEAssistantPythonFeature/O3DEAssistantPythonFeature.h [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+
+#include <AzCore/RTTI/RTTI.h>
+#include <AzCore/Serialization/SerializeContext.h>
+
+namespace GenAIFramework
+{
+    class O3DEAssistantPythonFeature : public PythonFeatureBase
+    {
+    public:
+        AZ_RTTI(O3DEAssistantPythonFeature, "{7c524a1b-a175-4c26-beb1-dbefb7f13f54}", PythonFeatureBase);
```
Comment (jhanca-robotecai) [inline Gems/GenAIFramework/Code/Source/O3DEAssistantPythonFeature/O3DEAssistantPythonFeature.h:23]: You might want to move the uuid to the common file with uuids.

File: Gems/GenAIFramework/Code/Source/Tools/GenAIAsyncRequestEditorSystemComponent.cpp [hunk 13]
```diff
@@ -7,6 +7,7 @@
  */
 
 #include "GenAIAsyncRequestEditorSystemComponent.h"
+#include <GenAIFramework/Feature/PythonFeatureBase.h>
```
Comment (jhanca-robotecai) [inline Gems/GenAIFramework/Code/Source/Tools/GenAIAsyncRequestEditorSystemComponent.cpp:10]: I guess this change is unnecessary. It creates another question: should we keep the Async part as it is?

File: Gems/GenAIFramework/Code/genaiframework_private_files.cmake [hunk 14]
```diff
@@ -17,4 +17,6 @@ set(FILES
 
     Source/O3DEAssistantFeature/O3DEAssistantFeature.cpp
     Source/O3DEAssistantFeature/O3DEAssistantFeature.h
+
+
```
Comment (jhanca-robotecai) [inline Gems/GenAIFramework/Code/genaiframework_private_files.cmake:21]:
```suggestion
```

File: Gems/GenAIFramework/Editor/Scripts/assistant.py [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+        self.handlerAgent.add_callback("OnAIResponse", self.OnAIResponse)
+
+    def OnNewMessage(self, message):
+        history = AIAgentRequestBus(bus.Event, "GetHistory", connectionId)
+
+        prompt = [{"role": "user", "content": [message[0]]}]
+        jsonHistory = json.loads(history)
+        fullPrompt = jsonHistory + prompt
+
+        fullJsonStringPrompt = json.dumps(fullPrompt)
```
Comment (jhanca-robotecai) [inline Gems/GenAIFramework/Editor/Scripts/assistant.py:52]: Don't you want to add a system message as done for the `cpp` implementation?

**Review by jhanca-robotecai — APPROVED** [review]
Great work, only minor comments from my side.

There are few tasks to solve as follow-ups:
- check if Async implementation for Python can be removed
- add a tutorial about creating a PythonFeature
- change the existing documentation to follow up the changes here (creating a cpp feature, information about inheritance and base classes)

I will create an issue for that.

Tested with Mistral, works correctly!

**Pull request merged**